File: src/main/java/co/uk/isxander/evergreenhud/elements/impl/ElementSpeed.java
/*
 * Copyright (C) Evergreen [2020 - 2021]
 * This program comes with ABSOLUTELY NO WARRANTY
 * This is free software, and you are welcome to redistribute it
 * under the certain conditions that can be found here
 * https://www.gnu.org/licenses/lgpl-3.0.en.html
 *
 * If you have any questions or concerns, please create
 * an issue on the github page that can be found here
 * https://github.com/Evergreen-Client/EvergreenHUD
 *
 * If you have a private concern, please contact
 * isXander @ business.isxander@gmail.com
 */

package co.uk.isxander.evergreenhud.elements.impl;

import co.uk.isxander.evergreenhud.elements.Element;
import co.uk.isxander.evergreenhud.settings.impl.EnumSetting;
import co.uk.isxander.evergreenhud.settings.impl.IntegerSetting;
import co.uk.isxander.evergreenhud.settings.impl.BooleanSetting;
import co.uk.isxander.evergreenhud.elements.ElementData;
import net.minecraft.util.MathHelper;
import net.minecraftforge.client.event.RenderGameOverlayEvent;

import java.text.DecimalFormat;

public class ElementSpeed extends Element {

    private double speed = 0;

    public IntegerSetting accuracy;
    public BooleanSetting trailingZeros;
    public BooleanSetting suffix;
    public EnumSetting<SpeedUnit> speedUnit;
    public BooleanSetting useX;
    public BooleanSetting useY;
    public BooleanSetting useZ;

    @Override
    public void initialise() {
        addSettings(speedUnit = new EnumSetting<>("Unit", "Functionality", "What unit of speed to display.", SpeedUnit.METERS_PER_SEC));
        addSettings(accuracy = new IntegerSetting("Accuracy", "Display", "How many decimal places the value should display.", 2, 0, 4, " places"));
        addSettings(trailingZeros = new BooleanSetting("Trailing Zeros", "Display", "Add zeroes to match the accuracy.", false));
        addSettings(suffix = new BooleanSetting("Suffix", "Display", "If the value should be suffixed with the specified unit.", false));
        addSettings(useX = new BooleanSetting("Use X", "Functionality", "Use the x coordinate when calculating the speed.", true));
        addSettings(useY = new BooleanSetting("Use Y", "Functionality", "Use the y coordinate when calculating the speed.", true));
        addSettings(useZ = new BooleanSetting("Use Z", "Functionality", "Use the z coordinate when calculating the speed.", true));
    }

    @Override
    public ElementData metadata() {
        return new ElementData("Speed Display", "Displays the blocks per second speed of the player.");
    }

    @Override
    protected String getValue() {
        String format = (trailingZeros.get() ? "0" : "#");
        StringBuilder sb = new StringBuilder(accuracy.get() < 1 ? format : format + ".");
        for (int i = 0; i < accuracy.get(); i++) sb.append(format);
        return new DecimalFormat(sb.toString()).format(speed) + (suffix.get() ? " " + speedUnit.get().display : "");
    }

    @Override
    public String getDefaultDisplayTitle() {
        return "Speed";
    }

    public void onRenderGameOverlay(RenderGameOverlayEvent.Post event) {
        if (event.type != RenderGameOverlayEvent.ElementType.ALL) return;

        double distTraveledLastTickX = (useX.get() ? mc.thePlayer.posX - mc.thePlayer.prevPosX : 0);
        double distTraveledLastTickY = (useY.get() ? mc.thePlayer.posY - mc.thePlayer.prevPosY : 0);
        double distTraveledLastTickZ = (useZ.get() ? mc.thePlayer.posZ - mc.thePlayer.prevPosZ : 0);

        this.speed = convertSpeed(MathHelper.sqrt_double(distTraveledLastTickX * distTraveledLastTickX + distTraveledLastTickY * distTraveledLastTickY + distTraveledLastTickZ * distTraveledLastTickZ) * 20);
    }

    private double convertSpeed(double speed) {
        switch (speedUnit.get()) {
            case MPH:
                speed *= 2.237;
                break;
            case KPH:
                speed *= 3.6;
                break;
        }
        return speed;
    }

    private enum SpeedUnit {
        METERS_PER_SEC("m/s"),
        MPH("mph"),
        KPH("kph");

        public final String display;
        SpeedUnit(String display) {
            this.display = display;
        }
    }

}

File: core/src/main/kotlin/dev/isxander/evergreenhud/elements/impl/ElementSpeed.kt
/*
 *
 * EvergreenHUD - A mod to improve on your heads-up-display.
 * Copyright (C) isXander [2019 - 2021]
 *
 * This program comes with ABSOLUTELY NO WARRANTY
 * This is free software, and you are welcome to redistribute it
 * under the certain conditions that can be found here
 * https://www.gnu.org/licenses/lgpl-2.1.en.html
 *
 * If you have any questions or concerns, please create
 * an issue on the github page that can be found here
 * https://github.com/isXander/EvergreenHUD
 *
 * If you have a private concern, please contact
 * isXander @ business.isxander@gmail.com
 *
 */

package dev.isxander.evergreenhud.elements.impl

import dev.isxander.evergreenhud.annotations.ElementMeta
import dev.isxander.evergreenhud.api.mc
import dev.isxander.evergreenhud.elements.type.SimpleTextElement
import dev.isxander.settxi.impl.OptionContainer
import dev.isxander.settxi.impl.boolean
import dev.isxander.settxi.impl.int
import dev.isxander.settxi.impl.option
import java.text.DecimalFormat
import kotlin.math.pow
import kotlin.math.sqrt

@ElementMeta(id = "SPEED", name = "Speed", category = "Player", description = "Display how fast you are moving.")
class ElementSpeed : SimpleTextElement() {
    var useX by boolean(
        default = true,
        name = "Use X",
        category = "Speed",
        description = "If the X coordinate is accounted for in the speed calculation.",
    )

    var useY by boolean(
        default = true,
        name = "Use Y",
        category = "Speed",
        description = "If the Y coordinate is accounted for in the speed calculation.",
    )

    var useZ by boolean(
        default = true,
        name = "Use Z",
        category = "Speed",
        description = "If the Z coordinate is accounted for in the speed calculation.",
    )

    var speedUnit by option(
        default = SpeedUnit.METERS_PER_SEC,
        name = "Speed Unit",
        category = "Speed",
        description = "In what unit should the speed be displayed in.",
    )

    var accuracy by int(
        default = 2,
        min = 0, max = 8,
        name = "Accuracy",
        category = "Speed",
        description = "How many decimal places to show the speed to.",
    )

    var trailingZeros by boolean(
        default = true,
        name = "Trailing Zeros",
        category = "Speed",
        description = "Show zeros to match the accuracy.",
    )

    var suffix by boolean(
        default = true,
        name = "Suffix",
        category = "Speed",
        description = "If the speed value is to be suffixed by your unit (e.g. m/s)",
    )

    init {
        title = "Speed"
    }

    override fun calculateValue(): String {
        var speed = 0.0

        if (!mc.player.isNull) {
            val distTraveledLastTickX = if (useX) mc.player.x - mc.player.prevX else 0.0
            val distTraveledLastTickY = if (useY) mc.player.y - mc.player.prevY else 0.0
            val distTraveledLastTickZ = if (useZ) mc.player.z - mc.player.prevZ else 0.0

            speed = convertSpeed(sqrt(distTraveledLastTickX.pow(2) + distTraveledLastTickY.pow(2) + distTraveledLastTickZ.pow(2)))
        }

        val format = if (trailingZeros) "0" else "#"
        var formattedSpeed =
            DecimalFormat("0${if (accuracy > 0) "." else ""}" + format.repeat(accuracy))
            .format(speed)

        if (suffix) formattedSpeed += " ${speedUnit.name}"
        return formattedSpeed
    }

    private fun convertSpeed(speed: Double): Double =
        when (speedUnit) {
            SpeedUnit.METERS_PER_SEC -> speed
            SpeedUnit.KPH -> speed * 3.6
            SpeedUnit.MPH -> speed * 2.237
            else -> throw IllegalStateException()
        }

    object SpeedUnit : OptionContainer() {
        val METERS_PER_SEC = option("m/s", "Meters per second")
        val KPH = option("kph", "Kilometers per hour")
        val MPH = option("mph", "Miles per hour")
    }
}
